avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerdeBase.java

```diff
@@ -136,6 +146,29 @@ protected Class compileClass(final String className, Set<String> knownUsedFullyQ
       throw new FastSerdeGeneratorException("Unable to compile:" + className + " from source file: " + filePath);
     }
 
-    return classLoader.loadClass(generatedPackageName + "." + className);
+    return loadClassWithLimit(() -> {
+      try {
+        return classLoader.loadClass(generatedPackageName + "." + className);
+      } catch (ClassNotFoundException e) {
+        throw new FastSerdeGeneratorException("Unable to load:" + className + " from source file: " + filePath, e);
+      }
+    });
+  }
+
+  /**
+   * A wrapper function that limits the total number of fast de/serializer classes loaded
+   *
+   * @param supplier The function to load fast de/serializer class
+   */
+  protected synchronized <T> T loadClassWithLimit(Supplier<T> supplier) {
+    if (this.loadClassNum < this.loadClassLimit) {
+      T classObj = null;
+      classObj = supplier.get();
+      this.loadClassNum++;
+      return classObj;
+    } else {
+      LOGGER.warn("Loaded fast serdes classes number {}, with limit set to {}", loadClassNum, loadClassLimit);
+    }
+    return null;
```

Comment:
I think it will be clearer to just throw the exception instead of  returning null.

Reply:
Sure, will fix it.